src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -769,11 +966,42 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
                             witness
                         }).collect()
                     } else {
-                        pats.into_iter().flat_map(|witness| {
-                            missing_ctors.iter().map(move |ctor| {
-                                witness.clone().push_wild_constructor(cx, ctor, pcx.ty)
-                            })
-                        }).collect()
+                        if consider_value_constructors {
+                            // If we've been trying to exhaustively match
+                            // over the domain of values for a type,
+                            // then we can provide better diagnostics
+                            // regarding which values were missing.
+                            missing_ctors.into_iter().map(|ctor| {
+                                match ctor {
+                                    // A constant range of length 1 is simply
+                                    // a constant value.
+                                    ConstantValue(value) => {
+                                        Witness(vec![Pattern {
+                                            ty: pcx.ty,
+                                            span: DUMMY_SP,
+                                            kind: box PatternKind::Constant { value },
+                                        }])
+                                    }
+                                    // We always report missing intervals
+                                    // in terms of inclusive ranges.
+                                    ConstantRange(lo, hi, end) => {
+                                        Witness(vec![Pattern {
+                                            ty: pcx.ty,
+                                            span: DUMMY_SP,
+                                            kind: box PatternKind::Range { lo, hi, end },
+                                        }])
+                                    },
+                                    _ => bug!("`ranges_subtract_pattern` should only produce \
+                                               `ConstantRange`s"),
+                                }
+                            }).collect()
+                        } else {
+                            pats.into_iter().flat_map(|witness| {
+                                missing_ctors.iter().map(move |ctor| {
+                                    witness.clone().push_wild_constructor(cx, ctor, pcx.ty)
```

Comment:
Shouldn't `push_wild_constructor` handle the generation of `PatternKind::{Constant,Range}` above?

Reply:
Yea the logic above should be moved to `push_wild_constructor`, as currently this code does not handle patterns like:
```Rust
match x {
    (1, _) => {}
    (_, None) => {}
}
```

Where the counterexample is `(0, Some(_))` + `(2.., Some(_))`. It would be good to add a test to this effect.

Also a comment:
```Rust
// extends the witness with a "wild" version of this constructor, that
// matches everything that can be built with it.
//
// For example, if `ctor` is a `Constructor::Variant` for `Option::Some`, this
// pushes the witness for `Some(_)`.
```

And a comment for, maybe, `Witness`:
```Rust
/// A witness of non-exhaustiveness for error reporting, represented
/// as a list of patterns with wildcards inside to represent elements
/// that can take any inhabitant of the type as a value.
///
/// A witness against a list of patterns should have the same types
/// and length as the pattern matched against. Because Rust `match`
/// is always against a single pattern, at the end the witness will
/// have length 1, but in the middle of the algorithm, it can contain
/// multiple patterns.
///
/// For example, if we are constructing a witness for the match against
/// ```
/// struct Pair(Option<(u32, u32)>, bool);
///
/// match (p: Pair) {
///    Pair(None, _) => {}
///    Pair(_, false) => {}
/// }
/// ```
///
/// We'll perform the following steps:
/// 1. Start with an empty witness
///     `Witness(vec![])`
/// 2. Push a witness `Some(_)` against the `None`
///     `Witness(vec![Some(_)])`
/// 3. Push a witness `true` against the `false`
///     `Witness(vec![Some(_), true])`
/// 4. Apply the `Pair` constructor to the witnesses
///     `Witness(vec![Pair(Some(_), true)])`
///
/// The final `Pair(Some(_), true)` is then the resulting witness.
#[derive(Clone, Debug)]
pub struct Witness<'tcx>(Vec<Pattern<'tcx>>);
```